libmscore/point.cpp

```diff
@@ -0,0 +1,252 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2018 Werner Schweer
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#include "point.h"
+
+#include "chord.h"
+#include "element.h"
+#include "measure.h"
+#include "mscore.h"
+#include "xml.h"
+
+namespace Ms {
+
+static constexpr PointInfo absDefaults = PointInfo::absolute();
+static constexpr PointInfo relDefaults = PointInfo::relative();
+
+//---------------------------------------------------------
+//   PointInfo::track
+//---------------------------------------------------------
+
+int PointInfo::track() const
+      {
+      if ((_staff == absDefaults._staff) || (_voice == absDefaults._voice))
+            return INT_MIN;
+      return VOICES * _staff + _voice;
+      }
+
+//---------------------------------------------------------
+//   PointInfo::setTrack
+//---------------------------------------------------------
+
+void PointInfo::setTrack(int track)
+      {
+      _staff = track / VOICES;
+      _voice = track % VOICES;
+      }
+
+//---------------------------------------------------------
+//   PointInfo::write
+//---------------------------------------------------------
+
+void PointInfo::write(XmlWriter& xml) const
+      {
+      if (isRelative()) {
+            xml.stag("move");
+            xml.tag("staves", _staff, relDefaults._staff);
+            xml.tag("voices", _voice, relDefaults._voice);
+            xml.tag("measures", _measure, relDefaults._measure);
+            xml.tag("fractions", _frac.reduced(), relDefaults._frac);
+            xml.tag("grace", _graceIndex, relDefaults._graceIndex);
+            xml.tag("notes", _note, relDefaults._note);
+            xml.etag();
+            }
+      else {
```

Comment:
No, in fact, it isn't used anywhere. Perhaps it is really better to remove it but I am not sure about that.

Reply:
I am happy for you to keep the ability for a PointInfo to be absolute *within the code* as that might come in handy, but I don't want to see `<move_abs>` in any MSCX files as it will spoil the diff. You could do something like `Q_ASSERT(isRelative())` in one of the save functions.

Reply:
OK, I'll remove that `<move_abs>`-related code then.